File: test-suite/src/test/java/io/micronaut/security/docs/authenticationparam/AuthenticationParamTest.java
package io.micronaut.security.docs.authenticationparam;

import io.micronaut.context.annotation.Property;
import io.micronaut.context.annotation.Requires;
import io.micronaut.core.type.Argument;
import io.micronaut.http.HttpRequest;
import io.micronaut.http.HttpResponse;
import io.micronaut.http.HttpStatus;
import io.micronaut.http.client.BlockingHttpClient;
import io.micronaut.http.client.HttpClient;
import io.micronaut.http.client.annotation.Client;
import io.micronaut.security.authentication.AuthenticationFailureReason;
import io.micronaut.security.authentication.AuthenticationRequest;
import io.micronaut.security.authentication.AuthenticationResponse;
import io.micronaut.security.authentication.provider.HttpRequestAuthenticationProvider;
import io.micronaut.test.extensions.junit5.annotation.MicronautTest;
import jakarta.inject.Singleton;
import org.junit.jupiter.api.Test;

import java.util.List;
import java.util.Map;

import static org.junit.jupiter.api.Assertions.assertEquals;
import static org.junit.jupiter.api.Assertions.assertFalse;
import static org.junit.jupiter.api.Assertions.assertTrue;

@Property(name = "spec.name", value = "AuthenticationParamTest")
@MicronautTest
class AuthenticationParamTest {

    @Test
    void authenticationCanBeUsedAsAControllerParameterToGetTheLoggedInUser(@Client("/") HttpClient httpClient) {
        BlockingHttpClient client = httpClient.toBlocking();
        HttpResponse<Map<String, Object>> rsp = client.exchange(HttpRequest.GET("/user/myinfo"), Argument.mapOf(String.class, Object.class));
        assertEquals(HttpStatus.OK, rsp.status());
        assertFalse(rsp.body().containsKey("username"));

        rsp = client.exchange(HttpRequest.GET("/user/myinfo").basicAuth("user", "password"), Argument.mapOf(String.class, Object.class));
        assertEquals(HttpStatus.OK, rsp.status());
        assertTrue(rsp.body().containsKey("username"));
        assertEquals("user", rsp.body().get("username"));
        assertEquals(List.of("ROLE_USER"), rsp.body().get("roles"));
    }

    @Requires(property = "spec.name", value = "AuthenticationParamTest")
    @Singleton
    static class AuthenticationProviderUserPassword<B> implements HttpRequestAuthenticationProvider<B> {
        @Override
        public AuthenticationResponse authenticate(HttpRequest<B> requestContext, AuthenticationRequest<String, String> authRequest) {
            return authRequest.getIdentity().equals("user") && authRequest.getSecret().equals("password")
                ? AuthenticationResponse.success("user", List.of("ROLE_USER"))
                : AuthenticationResponse.failure(AuthenticationFailureReason.CREDENTIALS_DO_NOT_MATCH);
        }
    }
}

File: test-suite-python/src/test/python/micronaut/security/docs/authenticationparam/AuthenticationParamTest.py
from typing import Annotated

from jakarta.inject import Inject, Singleton
from java.util import List, Map
from micronaut.context.annotation import Property, Requires
from micronaut.core.type import Argument
from micronaut.http import HttpRequest, HttpStatus
from micronaut.http.client import HttpClient
from micronaut.http.client.annotation import Client
from micronaut.security.authentication import AuthenticationFailureReason, AuthenticationRequest, AuthenticationResponse
from micronaut.security.authentication.provider import HttpRequestAuthenticationProvider
from micronaut.test.extensions.junit5.annotation import MicronautTest
from org.junit.jupiter.api import Test


@Property(name="spec.name", value="AuthenticationParamTest")
@MicronautTest
class AuthenticationParamTest:
    httpClient: Annotated[HttpClient, Inject, Client("/")]

    @Test
    def test_authentication_can_be_used_as_a_controller_parameter_to_get_the_logged_in_user(self):
        client = self.httpClient.toBlocking()
        rsp = client.exchange(HttpRequest.GET("/user/myinfo"), Argument.of(Map))
        assert rsp.status() == HttpStatus.OK
        assert not rsp.body().containsKey("username")

        rsp = client.exchange(HttpRequest.GET("/user/myinfo").basicAuth("user", "password"), Argument.of(Map))
        assert rsp.status() == HttpStatus.OK
        assert rsp.body().containsKey("username")
        assert rsp.body().get("username") == "user"
        assert list(rsp.body().get("roles")) == ["ROLE_USER"]


@Requires(property="spec.name", value="AuthenticationParamTest")
@Singleton
class AuthenticationProviderUserPassword(HttpRequestAuthenticationProvider):

    def authenticate(self, requestContext: HttpRequest, authRequest: AuthenticationRequest) -> AuthenticationResponse:
        if authRequest.getIdentity() == "user" and authRequest.getSecret() == "password":
            return AuthenticationResponse.success("user", List.of("ROLE_USER"))
        return AuthenticationResponse.failure(AuthenticationFailureReason.CREDENTIALS_DO_NOT_MATCH)
